File: src/lib/pulsenet/geo.ts
// Pure geo helpers — no DB access, so callers can batch-load countries once.

export function haversineKm(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const R = 6371
  const dLat = ((lat2 - lat1) * Math.PI) / 180
  const dLng = ((lng2 - lng1) * Math.PI) / 180
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLng / 2) ** 2
  return 2 * R * Math.asin(Math.sqrt(a))
}

export function nearestCountries(
  lat: number,
  lng: number,
  countries: { code: string; lat: number; lng: number }[],
  maxKm = 1500,
  limit = 2,
): string[] {
  return countries
    .map((c) => ({ code: c.code, d: haversineKm(lat, lng, c.lat, c.lng) }))
    .filter((x) => x.d <= maxKm)
    .sort((a, b) => a.d - b.d)
    .slice(0, limit)
    .map((x) => x.code)
}

/** Equirectangular projection of lat/lng onto a 0..1 x 0..1 plane (for SVG plotting). */
export function project(lat: number, lng: number): { x: number; y: number } {
  return { x: (lng + 180) / 360, y: (90 - lat) / 180 }
}

File: mini-services/pulsenet-engine/app/feeds/geo.py
"""Geo helpers — Haversine distance + nearest-country matching.

Mirrors src/lib/pulsenet/geo.ts so Python ingestion tags the same countries the
TS code would. Pure functions, no I/O.
"""

from __future__ import annotations

import math

EARTH_RADIUS_KM = 6371.0


def haversine_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Great-circle distance between two lat/lng points, in km."""
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlmb = math.radians(lng2 - lng1)
    a = math.sin(dphi / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dlmb / 2) ** 2
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(a))


def nearest_countries(
    lat: float,
    lng: float,
    countries: list[dict],
    max_km: float = 1500.0,
    limit: int = 3,
) -> list[str]:
    """Return ISO codes of the nearest catalog countries within max_km.

    Args:
        countries: list of dicts with keys code, lat, lng.
    """
    scored = []
    for c in countries:
        d = haversine_km(lat, lng, c["lat"], c["lng"])
        if d <= max_km:
            scored.append((d, c["code"]))
    scored.sort(key=lambda x: x[0])
    return [code for _, code in scored[:limit]]
